x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/search/InferencePhase.java

```diff
@@ -0,0 +1,128 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+package org.elasticsearch.xpack.ml.inference.search;
+
+import org.apache.lucene.util.SetOnce;
+import org.elasticsearch.action.ActionListener;
+import org.elasticsearch.action.LatchedActionListener;
+import org.elasticsearch.common.document.DocumentField;
+import org.elasticsearch.search.SearchExtBuilder;
+import org.elasticsearch.search.SearchHit;
+import org.elasticsearch.search.fetch.FetchSubPhase;
+import org.elasticsearch.search.internal.SearchContext;
+import org.elasticsearch.xpack.core.ml.inference.results.InferenceResults;
+import org.elasticsearch.xpack.core.ml.inference.trainedmodel.InferenceConfig;
+import org.elasticsearch.xpack.core.ml.utils.MapHelper;
+import org.elasticsearch.xpack.ml.inference.loadingservice.Model;
+import org.elasticsearch.xpack.ml.inference.loadingservice.ModelLoadingService;
+
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.CountDownLatch;
+
+/**
+ * A very rough sketch of a fetch sub phase that performs inference on each search hit,
+ * then augments the hit with the result.
+ */
+public class InferencePhase implements FetchSubPhase {
+
+    private final SetOnce<ModelLoadingService> modelLoadingService;
+
+    public InferencePhase(SetOnce<ModelLoadingService> modelLoadingService) {
+        this.modelLoadingService = modelLoadingService;
+    }
+
+    @Override
+    public void hitsExecute(SearchContext searchContext, SearchHit[] hits) {
+        SearchExtBuilder inferenceBuilder = searchContext.getSearchExt(InferenceSearchExtBuilder.NAME);
+        if (inferenceBuilder == null) {
+            return;
+        }
+
+        InferenceSearchExtBuilder infBuilder = (InferenceSearchExtBuilder)inferenceBuilder;
+
+        SetOnce<Model> model = new SetOnce<>();
+        CountDownLatch latch = new CountDownLatch(1);
+        ActionListener<Model> listener = new LatchedActionListener<>(
+                ActionListener.wrap(model::set, e -> { throw new RuntimeException();}), latch);
+
+        modelLoadingService.get().getModel(infBuilder.getModelId(), listener);
+        try {
+            // Eeek blocking on a latch we can't be doing that
```

Comment:
agreed :D. We may want this call to fail if the model is not deployed in the provided model service. Especially since there is no way to load it just in time :/